pyhf/pdf.py

```diff
@@ -126,7 +128,44 @@ def __init__(self, spec, **config_kwargs):
         # build up our representation of the specification
         self.config = _ModelConfig.from_spec(self.spec,**config_kwargs)
 
+    def _mtype_results(self,mtype,pars):
+        mtype_results = {}
+        for channel in self.spec['channels']:
+            for sample in channel['samples']:
+                for mname in sample['modifiers_by_type'].get(mtype,[]):
+                    modifier = self.config.modifier(mname)
+                    modpars  = pars[self.config.par_slice(mname)]
+                    mtype_results.setdefault(channel['name'],
+                            {}).setdefault(sample['name'],
+                            []).append(
+                                modifier.apply(channel, sample, modpars)
+                            )
+        return mtype_results
+
     def expected_sample(self, channel, sample, pars):
+        tensorlib, _ = get_backend()
+        #public API only, not efficient
+        all_modifications = self._all_modifications(pars)
+        return self._expected_sample(
+            tensorlib.astensor(sample['data']), #nominal
+            *all_modifications[channel['name']][sample['name']] #mods
+        )
+
+    def _expected_sample(self, nominal, factors, deltas):
+        tensorlib, _ = get_backend()
+        basefactor = [
+            tensorlib.sum(
+                tensorlib.stack(
+                    [nominal,tensorlib.sum(tensorlib.stack(deltas), axis=0)]),
+                axis=0)
+                if len(deltas) > 0 
+                else nominal
+        ]
+        factors += basefactor
+
+        return tensorlib.product(tensorlib.stack(tensorlib.simple_broadcast(*factors)), axis=0)
+
```

Comment:
this just implements how the modifications get combined with the nominal data. Up to now we have 2 types of modifications factors, and shifts (deltas) 

first we sum all deltas with the nominal and then apply all factors on that sum

Reply:
what this does is  compute

```
summands = [delta1,delta2,... nominal]
basefactor = np.sum(summands)
factors = [f1,f2,f3...,basefactor]
np.product(factors)

but in a follow up PR this entire function will be removed and `expected_actualdata` will look like this


```
        all_modifications = self._all_modifications(pars)
        delta_field  = np.zeros(self.cube.shape)
        factor_field = np.ones(self.cube.shape)
        for cname,smods in all_modifications.items():
            for sname,(factors,deltas) in smods.items():
                ind = self.hm[cname][sname]['index']
                for f in factors:
                    factor_field[ind] = factor_field[ind] * f
                for d in deltas:
                    delta_field[ind]  = delta_field[ind] + d
        combined = factor_field * (delta_field + self.cube)
```

Reply:
Hrmm, the way the code is currently written is verrrrry opaque to me. It looks like you sum up a bunch of summands, and then multiply those summand results together. Especially because you do `factors += basefactor`. Can you do sum and products separately, and return the combination of them rather than doing internal bookkeeping?